sound/soc/codecs/max98373-sdw.c

```diff
@@ -616,7 +616,11 @@ static int max98373_sdw_dai_hw_params(struct snd_pcm_substream *substream,
 
 	if (max98373->slot) {
 		stream_config.ch_count = max98373->slot;
-		port_config.ch_mask = max98373->rx_mask;
+
+		if (direction == SDW_DATA_DIR_RX)
+			port_config.ch_mask = max98373->rx_mask;
+		else
+			port_config.ch_mask = max98373->tx_mask;
```

Comment:
@plbossart Capture can work without tdm_slot for its ch_mask is alway 0x3 (I & V channel). But the ch count in hw_params is 4ch, so it can't be used for capture. And we also don't need tdm slot for playback since we have mirror mode support. Please check my patch.

Reply:
I can't follow your explanations,  too many negatives and contradicting arguments. Please re-explain what works and what does not.

Reply:
please check the latest PR. Now we don't need to set tdm in our machine driver.  For original code in capture case, params_channels(params) = 4, so we need tdm slot to constrain it to 2 for each codec.

Reply:
And for playback, we go with mirror mode and set mixer to choose input channel.

Reply:
For DSM usage:   we do with aplay -c 2 xxx.wav & arecord -c 4 xxx.wav.  In original codec for playback, ch =  2, mask = 0x3, the same as rx_mask in tdm setting, so we can skip it.  And for capture, we need set tx_mask = 0x3 to constrain ch_mask.